D. Increase the number of hidden layers

## Import the necessary libraries

In [1]:
import pandas as pd
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense, Input
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

## Load the concrete date from the CSV file

In [2]:
data = pd.read_csv('concrete_data.csv')

In [3]:
data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

## Define the datafrme of features X and target y

In [4]:
X = data.iloc[:, 0:8]
y = data.iloc[:, 8]

# Normalize the input features

Subtract the mean from the individual predictors and divide by the standard deviation.

In [5]:
X = X - X.mean() / X.std()

## Iterative training and evaluation of the neural network

In [6]:
def create_model():
  """
  Helper function to create a neural network using the given specifications
  """
  model = Sequential()
  # three hidden layer of 10 nodes and a ReLU activation function
  model.add(Dense(10, activation='relu', input_shape=(X.shape[1],)))
  model.add(Dense(10, activation='relu'))
  model.add(Dense(10, activation='relu'))
  # output layer
  model.add(Dense(1))
  # compile using the 'adam' optimizer and the 'mean squared error' as the loss function
  model.compile(optimizer='adam', loss='mean_squared_error')

  return model

In [7]:
mse_list = []

In [8]:
for _ in range(50):
  # Step 1: Randomly split the data into a training and test sets by holding 30% of the data for testing
  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

  # Step 2: Create the neural network model
  model = create_model()

  # Step 3: Train the model on the training data using 50 epochs
  model.fit(X_train, y_train, epochs=50, verbose=0)

  # Step 4: Evaluate the model on the test data
  y_pred = model.predict(X_test)

  # Step 5: Compute the mean squared error between the predicted concrete strength and the actual concrete strength.
  mse = mean_squared_error(y_test, y_pred)

  # Step 6: Insert the mean squared error in a list
  mse_list.append(mse)

10/10 [==============================] - 0s 3ms/step


In [9]:
# Step 7: Compute the mean and standard deviation of the mean squared errors
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

In [10]:
print(f"Mean of Mean Squared Errors: {mean_mse}")
print(f"Standard Deviation of Mean Squared Errors: {std_mse}")

Mean of Mean Squared Errors: 143.06484588405272
Standard Deviation of Mean Squared Errors: 153.03622034861905


The mean of MSE in Part B was `265.21468493109944`.

The mean of MSE in Part C was `194.93198887647642`.

In Part D, the mean of MSE is `143.06484588405272`.

The value of Mean of MSE of Part D is much smaller than that of Part B.
In the context of PART C, the smaller mean squared error suggests that our model is the best approximation of the best-fitting line.